Common/src/main/java/com/techarium/techarium/client/screen/MachineCoreScreen.java

```diff
@@ -0,0 +1,241 @@
+	private static final int TEXT_INPUT_Y = 26; // height 12
+
+	private final List<ResourceLocation> ids;
+	private final List<MultiBlockButton> buttons;
+	private EditBox textInput;
+	private int startIndex;
+	private boolean showHints;
+
+	public MachineCoreScreen(MachineCoreMenu menu, Inventory inventory, Component title) {
+		super(menu, inventory, title);
+		this.imageWidth = 165;
+		this.imageHeight = 186;
+		this.ids = this.menu.getAllMultiblockStructures();
+		this.buttons = new ArrayList<>();
+	}
+
+	@Override
+	protected void init() {
+		super.init();
+		this.addButtons("");
+		this.textInput = new EditBox(this.font, this.leftPos + TEXT_INPUT_X, this.topPos + TEXT_INPUT_Y, 100, 12, Component.empty());
+		this.addWidget(this.textInput);
+		this.setInitialFocus(this.textInput);
+		this.textInput.setFocus(true);
+		this.textInput.setBordered(false);
+		this.textInput.setResponder(this::filterIds);
+		this.showHints = false;
+	}
+
+	/**
+	 * Add selectable buttons for each multiblock.
+	 *
+	 * @param filter multiblock name filter (do nothing if empty).
+	 */
+	private void addButtons(String filter) {
+		this.startIndex = 0;
+		ResourceLocation selected = this.menu.selectedMultiblock();
+		ArrayList<MultiBlockButton> list = new ArrayList<>();
+		for (int i = 0; i < this.ids.size(); i++) {
+			ResourceLocation id = this.ids.get(i);
+			// name is created from resource location : techarium:random -> multiblock.techarium.random
+			MutableComponent name = Techarium.translatableComponent("multiblock." + id.getNamespace() + "." + id.getPath().replace("/", "."));
+			if (filter.isEmpty() || name.getString().toLowerCase().startsWith(filter.toLowerCase())) {
+				MultiBlockButton multiBlockButton = new MultiBlockButton(i, this.leftPos + IDS_X, 0, name, button -> {
+					this.buttons.forEach(element -> element.selected = false);
+					this.menu.setMultiBlock(id);
+					((MultiBlockButton) button).selected = true;
+				});
+				if (id.equals(selected)) {
+					multiBlockButton.selected = true;
+				}
+				list.add(multiBlockButton);
+			}
+		}
+		// sort the buttons by the translated name of the multiblocks
+		Collections.sort(list);
+		for (int i = 0; i < list.size(); i++) {
+			MultiBlockButton elem = list.get(i);
+			elem.y = this.topPos + IDS_Y + 14 * i;
+		}
+		this.buttons.addAll(list.stream().sorted().toList());
+		this.buttons.forEach(this::addRenderableWidget);
+		// hide buttons outside the area
+		for (int j = 8; j < this.buttons.size(); j++) {
+			this.buttons.get(j).visible = false;
+		}
+	}
+
+	private void filterIds(String filter) {
+		//remove every button
+		this.buttons.forEach(this::removeWidget);
+		this.buttons.clear();
+		//add back buttons starting with filter
+		this.addButtons(filter);
+	}
+
+	@Override
+	public boolean keyPressed(int keyCode, int scanCode, int modifiers) {
+		if (keyCode == GLFW.GLFW_KEY_E && this.textInput.isFocused()) {
+			// I don't know why, but the screen is closed even if the text input is focused
```

Comment:
It seems that charTyped which text boxes use is called after keyPressed so do what the anvil does so it can be better handled for more situations because `!isFocused()` is not the only state where closing the screen is permitted when there is an editbox.